### PR TITLE
Sanitize PDF embeds

### LINKED ISSUE #6083: Add support for embedding PDFs
## Feature description

Gutenberg 10.5 adds support for embedding PDFs in the File block: https://github.com/WordPress/gutenberg/pull/30857.

It does so by adding an [`object`](https://github.com/WordPress/gutenberg/blob/d504910ad11339dff0d3186db32bc39f3c665501/packages/block-library/src/file/save.js#L32-L41) element to the page:

```jsx
<object
	className="wp-block-file__embed"
	data={ href }
	type="application/pdf"
	style={ {
		width: '100%',
		height: `${ previewHeight }px`,
	} }
	aria-label={ pdfEmbedLabel }
/>
```

We need to add an `AMP_Object_Sanitizer` which automatically converts `object[type=application/pdf]` into [`amp-google-document-embed`](https://amp.dev/documentation/components/amp-google-document-embed/?format=websites).

We also need to dequeue the new `wp-block-library-file` script which is getting enqueued in the `render_callback`, so dequeuing will need to be done at the `wp_print_footer_scripts` action. Or otherwise, we can just unset the `render_callback` entirely so that the `$content` is just passed through as-is.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by pierlon**
## Summary

Fixes #6083.

(Note that being unable to see the embed on the AMP page is expected if the PDF is hosted locally.)

## Todo

- [x] Determine why the `amp-google-document-embed` component is not being rendered in the Twenty Twenty theme.

## Demo

image

## QA Instructions

1. Create a new post
2. Insert a file block into the editor
3. Upload a PDF document
4. Verify that the PDF document is displayed on the AMP page

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: includes/embeds/class-amp-core-block-handler.php [hunk 1]
```diff
@@ -211,6 +212,41 @@ public function ampify_video_block( $block_content, $block ) {
 		return $block_content;
 	}
 
+	/**
+	 * Ampify file block.
+	 *
+	 * Inject the an inline style attribute to handle PDF embeds not being displayed in some themes (eg. Twenty Twenty).
```
Comment (schlessera) [inline includes/embeds/class-amp-core-block-handler.php]:
```suggestion
	 * Inject an inline style attribute to handle PDF embeds not being displayed in some themes (eg. Twenty Twenty).
```

File: includes/embeds/class-amp-core-block-handler.php [hunk 2]
```diff
@@ -211,6 +212,41 @@ public function ampify_video_block( $block_content, $block ) {
...
+	 * @return string Filtered block content.
+	 */
+	public function ampify_file_block( $block_content, $block ) {
+		if ( empty( $block['attrs']['displayPreview'] ) || '.pdf' !== substr( $block['attrs']['href'], -4 ) ) {
+			return $block_content;
+		}
+
+		add_action(
+			'wp_print_footer_scripts',
+			static function() {
```
Comment (schlessera) [inline includes/embeds/class-amp-core-block-handler.php]:
```suggestion
			static function () {
```

File: includes/sanitizers/class-amp-object-sanitizer.php [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+use AmpProject\Layout;
+
+/**
+ * Class AMP_Object_Sanitizer
+ *
+ * Sanitizes `<object>` embeds.
+ *
+ * @since 2.1
+ * @internal
+ */
```
Comment (schlessera) [inline includes/sanitizers/class-amp-object-sanitizer.php]: Just as a note for now: the file and class docblocks for the sanitizers need some general overhaul. I'm not marking anything here to fix this for this PR, as the above is consistent with the rest. But the current way of documenting these:

- Fails to take all WPCS rules into account.
- Has value-less content that just repeats the code.
- Has the wrong package annotation.

File: includes/sanitizers/class-amp-object-sanitizer.php [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	/**
+	 * Sanitize PDF embeds.
+	 *
+	 * @see \AMP_Core_Block_Handler::ampify_file_block()
+	 *
+	 * @param DOMElement $element Object element.
+	 */
+	public function sanitize_pdf( DOMElement $element ) {
```
Comment (schlessera) [inline includes/sanitizers/class-amp-object-sanitizer.php]: The code can now make use of the extended `AmpProject\Dom\Element`.
```suggestion
	 * @param Element $element Object element.
	 */
	public function sanitize_pdf( Element $element ) {
```

File: includes/sanitizers/class-amp-object-sanitizer.php [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+		];
+
+		$title = $element->getAttribute( 'aria-label' );
+		if ( '' !== $title ) {
+			$attributes[ Attribute::TITLE ] = $title;
+		}
+
+		// If it is a block, retain original styling for element.
+		if (
+			$element->parentNode instanceof DOMElement
```
Comment (schlessera) [inline includes/sanitizers/class-amp-object-sanitizer.php]:
```suggestion
			$element->parentNode instanceof Element
```

File: includes/sanitizers/class-amp-object-sanitizer.php [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	 *
+	 * @var int
+	 */
+	const DEFAULT_PDF_EMBED_HEIGHT = 600;
+
+	/**
+	 * Sanitize `object` elements from the HTML contained in this instance's Dom\Document.
+	 */
+	public function sanitize() {
+		$elements = $this->dom->getElementsByTagName( 'object' );
```
Comment (schlessera) [inline includes/sanitizers/class-amp-object-sanitizer.php]: Requires `composer update` for https://github.com/ampproject/amp-toolbox-php/pull/169/files?diff=unified&w=1
```suggestion
		$elements = $this->dom->getElementsByTagName( Tag::OBJECT );
```

File: includes/sanitizers/class-amp-object-sanitizer.php [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+		if ( 0 === $elements->length ) {
+			return;
+		}
+
+		/** @var DOMElement $element */
+		foreach ( $elements as $element ) {
+			if ( $element->getAttribute( Attribute::TYPE ) === 'application/pdf' ) {
```
Comment (schlessera) [inline includes/sanitizers/class-amp-object-sanitizer.php]: Requires `composer update` for https://github.com/ampproject/amp-toolbox-php/pull/169/files?diff=unified&w=1
```suggestion
			if ( $element->getAttribute( Attribute::TYPE ) === Attribute::TYPE_PDF ) {
```

File: includes/sanitizers/class-amp-object-sanitizer.php [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+	 * @param DOMElement $element Object element.
+	 */
+	public function sanitize_pdf( DOMElement $element ) {
+		$parsed_style = $this->parse_style_string( $element->getAttribute( Attribute::STYLE ) );
+		$embed_height = isset( $parsed_style['height'] ) ? $parsed_style['height'] : self::DEFAULT_PDF_EMBED_HEIGHT;
+
+		$attributes = [
+			Attribute::LAYOUT => Layout::FIXED_HEIGHT,
+			Attribute::HEIGHT => $embed_height,
+			Attribute::SRC    => $element->getAttribute( 'data' ),
```
Comment (schlessera) [inline includes/sanitizers/class-amp-object-sanitizer.php]: Requires `composer update` for https://github.com/ampproject/amp-toolbox-php/pull/169/files?diff=unified&w=1
```suggestion
			Attribute::SRC    => $element->getAttribute( Attribute::DATA ),
```

File: includes/sanitizers/class-amp-object-sanitizer.php [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+		];
+
+		$title = $element->getAttribute( 'aria-label' );
```
Comment (schlessera) [inline includes/sanitizers/class-amp-object-sanitizer.php]: Requires `composer update` for https://github.com/ampproject/amp-toolbox-php/pull/169/files?diff=unified&w=1
```suggestion
		$title = $element->getAttribute( Attribute::ARIA_LABEL );
```

File: includes/sanitizers/class-amp-object-sanitizer.php [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+			in_array(
+				'wp-block-file',
+				explode( ' ', $element->parentNode->getAttribute( Attribute::CLASS_ ) ),
+				true
+			)
+		) {
+			$attributes[ Attribute::CLASS_ ] = 'wp-block-file__embed';
+		}
+
+		$amp_element = AMP_DOM_Utils::create_node( $element->ownerDocument, 'amp-google-document-embed', $attributes );
```
Comment (schlessera) [inline includes/sanitizers/class-amp-object-sanitizer.php]: Requires `composer update` for https://github.com/ampproject/amp-toolbox-php/pull/169/files?diff=unified&w=1
```suggestion
		$amp_element = AMP_DOM_Utils::create_node( $element->ownerDocument, Extension::GOOGLE_DOCUMENT_EMBED, $attributes );
```

File: includes/sanitizers/class-amp-object-sanitizer.php [hunk 11]
```diff
@@ -0,0 +1,84 @@
+<?php
+/**
+ * Class AMP_Object_Sanitizer
+ *
+ * @package AMP
+ */
+
+use AmpProject\Attribute;
+use AmpProject\Layout;
```
Comment (schlessera) [inline includes/sanitizers/class-amp-object-sanitizer.php]:
```suggestion
use AmpProject\Attribute;
use AmpProject\Dom\Element;
use AmpProject\Extension;
use AmpProject\Layout;
use AmpProject\Tag;
```

**Review by schlessera — CHANGES_REQUESTED** [review]
I pushed an update to the toolbox to add all missing constants for this sanitizer: https://github.com/ampproject/amp-toolbox-php/pull/169/files?diff=unified&w=1

Note that those kinds of updates will be almost never needed once https://github.com/ampproject/amp-toolbox-php/pull/100 was merged.

**Comment by pierlon** [comment]
Thanks for the review @schlessera. I've made use of those new constants in https://github.com/ampproject/amp-wp/pull/6112/commits/55a11f72d7d304b552caafa1aa395b89cb758af4.

**Review by schlessera — APPROVED** [review]

File: includes/embeds/class-amp-core-block-handler.php [hunk 12]
```diff
@@ -211,6 +212,41 @@ public function ampify_video_block( $block_content, $block ) {
...
+			0
+		);
+
+		// In Twenty Twenty the PDF embed fails to render due to the parent of the embed having
+		// the style rule `display: flex`. Overriding the display to `block` seems to do the trick.
+		return preg_replace(
+			'/(?<=<div\s)[^>]*class=(?:"|"[^"]*\s)wp-block-file(?:"|\s[^"]*")[^>]*>/',
+			'style="display: block;" $0',
+			$block_content
+		);
```
Comment (westonruter) [inline includes/embeds/class-amp-core-block-handler.php]: This appears to not not be ideal as the flexbox was used to make the Download button appear on the right, which now doesn

Non-AMP | AMP
-----------|------
image | image

It seems the more fundamental issue is that the `amp-google-document-embed` doesn't have `width:100%`. I'm gonna try that approach instead.

Comment (westonruter) [reply includes/embeds/class-amp-core-block-handler.php]: There, with 7cef7538e2583024fef4805309957de9cf34f8f5:

Non-AMP | AMP
-----------|------
image | image
image | image

**Review by westonruter — APPROVED** [review]

**Comment by pierlon** [comment]
Thanks for the review @westonruter, changes are :ok_hand:.

**Pull request merged**